### ISSUE TITLE
Add ability to preload fonts

### ISSUE THREAD

**Issue by osaxma** [material-foundation/flutter-packages#151]
**Issue**

When a Google font is used the first time, the text shows in the fallback font initially, and then shows the actual Google Font as shown in the attached recording.

This behavior appears in both cases:
- Font is available in assets and `GoogleFonts.config.allowRuntimeFetching = false`.
- Font is fetched from network/cache and not available in asset. 

The recording is based on the second case in a Flutter Web app, and the font is most likely coming from cache. For the first case, the font-switch is quicker but it is still apparent.

https://user-images.githubusercontent.com/46427323/109412438-0da8a980-79b9-11eb-9779-e098d210985d.mov

**Request**
It'd be nice to have an asynchronous function that can be called and waited for to preload the fonts for a given list of fonts or from all the fonts in the assets. This will allow avoiding this undesirable font-switch from showing up.  An example API for the suggested request would be something like this: 

```
final fonts = <String>['Caveat', 'Roboto', ....... , 'Lato'];
await GoogleFonts.preloadFonts(fonts);

// or
await GoogleFonts.preloadAssetFonts();
``` 

**Additional Context**

The following was previously suggested by @clocksmith in #103 but the issue is closed:
>When fonts are coming from assets, they still need to be loaded by the FontLoader, which is very fast when loading from assets, but not necessarily done before the first screen. We might need to make a preload function available to support this case (something that calls loadFontIfNecessary).

**Comment by marchellodev** [comment]
+ Would be very useful. Came here to create exactly the same issue.

**Comment by dJani97** [comment]
I came to report the same. +1 👍

**Comment by clocksmith** [comment]
For this solution, would be it ok to block on the splash screen while the fonts are loaded, causing a slightly delayed start up time?

**Comment by dJani97** [comment]
@clocksmith I think that would be the best.

**Comment by marchellodev** [comment]
No. This way if the app supports multiple languages, the splash screen will be dramatically delayed, even though the user does not necessarily needs all those fonts for multiple languages.

It would be much better if we had an async function using which we could preload specific fonts.

**Comment by TissueFluid** [comment]
I also found this issue when drawing some text to the canvas. And since the font was not loaded when the canvas was drawn, the font remains unchanged. Unless I triggered repainting the canvas by for example increase the font size.

**Comment by spydon** [comment]
> I also found this issue when drawing some text to the canvas. And since the font was not loaded when the canvas was drawn, the font remains unchanged. Unless I triggered repainting the canvas, it shows the font properly.

This is the same issue as we are facing with google fonts in [Flame](https://github.com/flame-engine/flame) too, since we are drawing the text directly on the canvas.

**Comment by npsulav** [comment]
Wanted to create a same issue. Its a mandatory feature.

**Comment by mrcndn** [comment]
+1 👍

**Comment by duck-dev-go** [comment]
> > I also found this issue when drawing some text to the canvas. And since the font was not loaded when the canvas was drawn, the font remains unchanged. Unless I triggered repainting the canvas, it shows the font properly.
> 
> This is the same issue as we are facing with google fonts in [Flame](https://github.com/flame-engine/flame) too, since we are drawing the text directly on the canvas.

Also having issues using it in flame because it is not preloaded. But in our usecase we are not loading from assets and we are loading the fonts dynamically. It would be nice to await preloading specific fonts right before they are we get them.

Something like

`await GoogleFonts.getFontAsync()`

**Comment by duck-dev-go** [comment]
Made a feature request [here](https://github.com/material-foundation/google-fonts-flutter/pull/273)

@spydon seems to be working with flame components also and it is possible to make multiple calls from multiple components to preload a font.

**Comment by mugi-luffy** [comment]
same issue 
+1

**Comment by maRci002** [comment]
Here is some dirty hack to listen when new font is ready to use.

This solution is based on [catching events from the engine](https://github.com/flutter/engine/blob/afdc32933be772d01503072fc33c8bf19f09abf0/lib/ui/text.dart#L3243-L3260).
It is strongly discouraged to use this in a production environment, as if the font cannot be loaded, the event will not fire. Moreover registering an existing listener to `ChannelBuffers` results to lose the original one so it's better to use `PlatformDispatcher` in order to use it just change `if (oldOnPlatformMessage != null) {` to `if (true) {` and `oldOnPlatformMessage(name, data, callback);` to `oldOnPlatformMessage?.call(name, data, callback);`.

```dart
import 'dart:async';
import 'dart:convert';
import 'dart:ui';

import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:google_fonts/google_fonts.dart';

void main() async {
  WidgetsFlutterBinding.ensureInitialized();
  GoogleFonts.lato();
  await fontLoadStream(fontsToLoad: 1).last;

  runApp(const MyApp());
}

final _fontChangeMessage = utf8.encoder.convert(json.encode(<String, Object?>{'type': 'fontsChange'})).buffer.asByteData();

Stream<void> fontLoadStream({int? fontsToLoad}) {
  late StreamController<void> controller;
  var loadedFonts = 0;

  void onData(ByteData? data) {
    if (data != null) {
      if (data.lengthInBytes == _fontChangeMessage.lengthInBytes && listEquals(data.buffer.asUint8List(), _fontChangeMessage.buffer.asUint8List())) {
        loadedFonts++;

        print('Fonts loaded: $loadedFonts');
        controller.add(null);

        if (loadedFonts == fontsToLoad) {
          controller.close();
        }
      }
    }
  }

  const kSystemChannelName = 'flutter/system';
  void channelBuffersListener(ByteData? data, PlatformMessageResponseCallback callback) {
    onData(data);
  }

  final oldOnPlatformMessage = PlatformDispatcher.instance.onPlatformMessage;

  void addListener() {
    if (oldOnPlatformMessage != null) {
      PlatformDispatcher.instance.onPlatformMessage = (name, ByteData? data, callback) {
        if (name == kSystemChannelName) {
          onData(data);
        }
        oldOnPlatformMessage(name, data, callback);
      };
    } else {
      channelBuffers.setListener(kSystemChannelName, channelBuffersListener);
    }
  }

  void removeListener() {
    PlatformDispatcher.instance.onPlatformMessage = oldOnPlatformMessage;
    channelBuffers.clearListener(kSystemChannelName);
  }

  controller = StreamController<void>(
    onListen: addListener,
    onPause: removeListener,
    onResume: addListener,
    onCancel: removeListener,
  );

  return controller.stream;
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        body: Center(
          child: MyWidget(),
        ),
      ),
    );
  }
}

class MyWidget extends StatelessWidget {
  const MyWidget({super.key});

  @override
  Widget build(BuildContext context) {
    return Text(
      'Hello, World!',
      style: GoogleFonts.lato(),
    );
  }
}
```

**Comment by guidezpl** [comment]
Update: this will be done this quarter

Googlers, see go/preloading-fonts-flutter

**Comment by jtkeyva** [comment]
@guidezpl
quarter is almost up! haha j/k couldn't resist the saying. hope to see this implemented soon.
thanks

**Comment by MiniSuperDev** [comment]
Hello, Is there any way to do it while it is being implemented?

For example is there a way to know if a font was loaded in the [FontLoader](https://api.flutter.dev/flutter/services/FontLoader-class.html)?
I want to know if a google font is already loaded, just checking if the font family already exists, if not I know I should loaded and I wait a delay until is loaded.

Thank you

**Comment by osaxma** [comment]
🎉 🎊 🎈

**Comment by RicardoMudinyane** [comment]
This is still not done ??

**Comment by guidezpl** [comment]
This is a freely available package I work on when I have time. If you are unsatisfied, I encourage you to contribute. Beyond that, I closed this with #195.
